File: bsplines/python/dos_electrones/1D_efectivo/vs_B/cl2e.py
# Codigo que calcula los autovalores y autovectores del hamiltoniano
# efectio 1D para un electron.
#
# El hamiltoniano del sistema es
#
#   Hz = -hbar^2/(2*me)*d^2/dz^2 + V_{ef}(z)
#
# El potencial de confinamiento efectivo es
#
#   V_{ef}(z) = (v1*l^2/(1+l^2/(2*sigma^2))^2 + (v1*z^2-v2)/(1+l^2/(2*sigma^2)))*exp(-z^2/(2*sigma^2))
#
# La matriz del hamiltoniano la calculo usando como base
# del espacio los B-splines. El problema de autovalores es un
# problema generalizado
#
#     H*c = E*S*c
#
# Guardo, los autovalores del hamiltoniano en funcion del
# campo magnetico aplicado.
#######################################################################
#######################################################################
import numpy as np
import matplotlib.pyplot as plt
import scipy.sparse.linalg
from scipy import linalg as LA
from bspline import Bspline
from sys import exit
from knots import knots_sequence
from index import indices
from interaccion import V_interaccion
from simetrizacion import H_antisimetrico

## Defino el potencial del pozo en la coordenada z
def V_potencial_pozo_z1(z1, x):
	U = np.piecewise(x, [x<-z1, (-z1<=x)&(x<z1), z1<x], [0, 1, 0]);
	return U

def V_potencial_pozo_z2(z1, z2, x):
	U = np.piecewise(x, [x<-z2, (-z2<=x)&(x<-z1), (-z1<=x)&(x<=z1), (z1<x)&(x<=z2), z2<x], [0, 1, 0, 1, 0]);
	return U

#######################################
#######################################
## Aca empieza el main del programa ###
#######################################
#######################################

## Parametros utiles
a0 = 0.0529177210; eV = 27.21138564; c_light = 137.035999074492; ua_to_T = 1.72e3;
alpha = 658.4092645439;

## Parametros fisicos del problema
me		 = 0.041; ## Masa de la particula
r0		 = 7.0; ## Ancho del pozo gaussiano en nm
az		 = 7.0;
bz		 = 2.5;
v1		 = 0.37; ## Alto de la barrera
v2		 = 0.108844; ## Profundidad del pozo
delta	 = 0.1;
eta		 = 0.1;
B_i		 = 1.0;
B_f 	 = 50.0;

bcampo_vec = np.linspace(B_i, B_f, 49);

# Pendiente de los niveles de landau
slope_ll = eV/(me*c_light*ua_to_T);

## Intervalo de integracion en la coordenada z
Zmax = 1000.0;
Zmin = -Zmax;

N_intervalos_z = 35;
N_cuad = 10;
grado = 6;
kord = grado + 1;
beta = 0.0065; ## Cte de decaimiento para los knots en la distribucion exponencial
N_splines_z = N_intervalos_z + grado;
N_base_z = N_splines_z - 2;

N_dim = N_base_z; # Tamano de las matrices

nev = N_dim-1;

if nev > N_dim:
	print "nev > N_dim, cambiar el valor de nev"
	exit(0)

archivo1 = "./resultados/E_vs_B-v1_%6.4feV-v2_%6.4feV-az_%6.4f-bz_%6.4f-eta_%6.4f.dat" % (v1, v2, az, bz, eta)

f1 = open(archivo1, 'w')
f1.write("# Intervalo de integracion en z [{0}, {1}]\n".format(Zmin, Zmax))
f1.write("# Grado de los B-splines {0}\n".format(grado))
f1.write("# Num de intervalos {0} y tamano de base {1} en z\n".format(N_intervalos_z, N_base_z))
f1.write("# Dimension total del espacio N_dim = N_base_z = {0}\n".format(N_dim))
f1.write("# Cte de separacion de knots en dist exp beta = {0}\n".format(beta))
f1.write("# Orden de la cuadratura N_cuad = {0}\n".format(N_cuad))
f1.write("# Masa de la particula me = {0} en UA\n".format(me))
f1.write("# Parametro del pozo az = {0} en nm\n".format(az))
f1.write("# Parametro del pozo bz = {0} en nm\n".format(bz))
f1.write("# Parametro del pozo r0 = {0} en nm\n".format(r0))
f1.write("# Parametro del pozo v1 = {0} en eV\n".format(v1))
f1.write("# Parametro del pozo v2 = {0} en eV\n".format(v2))
f1.write("# Parametro de normalizacion de coulomb 1D delta = {0} en nm\n".format(delta))
f1.write("# Carga efectiva de la interaccion eta = {0} en UA\n".format(eta))
f1.write("# Autovalores calculados\n")

## Paso a unidades atomicas
Zmax = Zmax/a0; Zmin = Zmin/a0;
beta = beta*a0;
az = az/a0; bz = bz/a0; r0 = r0/a0; delta = delta/a0;
v1 = v1/eV; v2 = v2/eV;

## Vector de knots para definir los B-splines, distribucion uniforme
knots_z = knots_sequence(grado, 'exp', N_intervalos_z, beta, Zmin, Zmax);

## Pesos y nodos para la cuadratura de Gauss-Hermite
x, w = np.polynomial.legendre.leggauss(N_cuad);

## Escribo bien los nodos y los pesos para las integrales en z
z_nodos = np.array([]);
w_pesos = np.array([]);
for i in range(N_intervalos_z):
	aux_x = 0.5*(knots_z[i+grado+1]-knots_z[i+grado])*x + 0.5*(knots_z[i+grado+1]+knots_z[i+grado]);
	aux_w = 0.5*(knots_z[i+grado+1]-knots_z[i+grado])*w;

	z_nodos = np.hstack((z_nodos, aux_x));
	w_pesos = np.hstack((w_pesos, aux_w));

w_pesos2 = np.tile(w_pesos, (N_base_z, 1));
z_nodos2 = np.tile(z_nodos, (N_base_z, 1));

## B-splines en la coordenada z
basis = Bspline(knots_z, grado);
## Calculo la matriz con los bsplines y sus derivadas en z
bsz  = [basis._Bspline__basis(i, basis.p) for i in z_nodos]; # evaluo los bsplines
dbsz = [basis.d(i) for i in z_nodos];                        # evaluo las derivadas de los bsplines

bsz = np.array(bsz)
bsz = bsz[:,1:N_splines_z-1]

dbsz = np.array(dbsz)
dbsz = dbsz[:,1:N_splines_z-1]


# Llamo la funcion que calcula la matriz de la interaccion
V_int = V_interaccion(z_nodos, w_pesos, delta, bsz)


### Ahora paso a armar las matrices para el hamiltoniano de una particula
z_mat = np.dot(np.transpose(bsz), (np.transpose(z_nodos2*w_pesos2)*bsz));

## Matriz de solapamiento en z
Sz = np.dot(np.transpose(bsz), (np.transpose(w_pesos2)*bsz));

## Matriz de energia cinetica en z
Tz = 0.5/me*np.dot(np.transpose(dbsz), (np.transpose(w_pesos2)*dbsz));

## Matriz del potencial de confinamiento
Uz = V_potencial_pozo_z1(0.5*az, z_nodos)
Uz = np.tile(Uz, (N_base_z, 1))

Vz1 = np.dot(np.transpose(bsz), (np.transpose(Uz*w_pesos2)*bsz));

Uz = V_potencial_pozo_z2(0.5*az, 0.5*(az+bz), z_nodos)
Uz = np.tile(Uz, (N_base_z, 1))

Vz2 = np.dot(np.transpose(bsz), (np.transpose(Uz*w_pesos2)*bsz));

Tr = 0.5/me*9.74494e-5
Ur = 0.5*me*10677.5
Vr = 0.82604193

## Calculo los autovalores y autovectores
# auval = np.zeros(N_dim);
for bcampo in bcampo_vec:

	l_campo = np.sqrt(alpha/bcampo)/a0;
	omega = 0.5*bcampo/(me*c_light*ua_to_T);  ## Frecuencia de oscilacion debida al campo

	Hr = Tr + omega**2*Ur

	# El hamiltoniano es
	Hz = Tz + v1*Vz2 - v2*Vr*Vz1 + Hr*Sz;

	e1, auvec = LA.eigh(Hz, Sz);
	e1 = eV*e1;

	H_sim, S_sim = H_antisimetrico(Hz, Sz, V_int, eta)

	e, auvec = LA.eig(H_sim, S_sim);
	# e, auvec = scipy.sparse.linalg.eigsh(H_sim, k=np.size(H_sim[0,:])/2, M=S_sim, sigma=None, which='SM')
	e = eV*e;

	# auval = np.vstack((auval, e));

	f1.write("{:13.9e}   ".format(bcampo))
	for i in range(nev):
		f1.write("{:13.9e}   ".format(e.real[i]))

	f1.write("\n")

# for i in range(5):
# 	plt.plot(bcampo_vec, auval[1:,i])
#
# plt.show()
f1.close()

File: bsplines/fortran/dos_electrones/1D_efectivo/pozo_cte_radial_infinito/sol_landau/vs_B/coulomb_rectificado/cl2e.f90

!!!!!!!!!!!!!!!!!!!!
#ifndef ZMIN_
#define ZMIN_  0._pr
#endif

#ifndef ZMAX_
#define ZMAX_ 10._pr
#endif

#ifndef TIPO_
#define TIPO_ 1
#endif

#ifndef BETA_
#define BETA_ 0._pr
#endif

#ifndef KORD_
#define KORD_ 5
#endif

#ifndef L_INTERVAL_
#define L_INTERVAL_ 30
#endif

#ifndef N_CUAD_
#define N_CUAD_ 100
#endif

#ifndef NEV_
#define NEV_ 10
#endif

#ifndef ME_
#define ME_ 1._pr
#endif

#ifndef AZ_
#define AZ_ 2._pr
#endif

#ifndef BZ_
#define BZ_ 5._pr
#endif

#ifndef R0_
#define R0_ 2._pr
#endif

#ifndef RMAX_
#define RMAX_ 50._pr
#endif

#ifndef V1_
#define V1_ 0._pr
#endif

#ifndef V2_
#define V2_ 1._pr
#endif

#ifndef ETA_
#define ETA_ 1._pr
#endif

#ifndef DELTA_
#define DELTA_ 0.5_pr
#endif

#ifndef BCAMPO_I_
#define BCAMPO_I_ 1._pr
#endif

#ifndef BCAMPO_F_
#define BCAMPO_F_ 10._pr
#endif

#ifndef NUM_PUNTOS_B_
#define NUM_PUNTOS_B_ 10
#endif
!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!
program main
  use precision, pr => dp
  use cuadratura_knots
  use matrices
  use mod_eig
  implicit none
  integer :: tipo, kord, l_interval, n_cuad, nev
  integer :: num_puntos_b
  real(pr) :: zmin, zmax, beta, me, v1, v2
  real(pr) :: az, bz, r0, rmax, delta, eta
  real(pr) :: bcampo_i, bcampo_f
  real(pr) :: vr, vrmax, omega, slope_ll
  real(pr), parameter :: a0 = 0.0529177210_pr, eV = 27.21138564_pr
  real(pr), parameter :: alpha = 658.4092645439_pr, c_light = 137.035999074492_pr
  real(pr), parameter :: ua_to_T = 1.72d3
  !!!!!!
  integer :: i, nk, nb, ndimh, ind_b
  integer, allocatable :: k(:)
  real(pr) :: delta_b, bcampo, l_campo
  real(pr), allocatable :: t(:)
  real(pr), allocatable :: x(:,:), w(:,:), pl(:,:)
  real(pr), allocatable :: s(:,:), v01(:,:), v02(:,:), ke(:,:)
  real(pr), allocatable :: V_int(:,:,:,:)
  real(pr), allocatable :: h(:,:), ms(:,:), mv(:,:), auval(:), auvec(:,:)
  real(pr), allocatable :: exp_int(:,:)
  character(150) :: file_auval, file_exp

  zmin = real(ZMIN_, pr); zmax = real(ZMAX_, pr);
  tipo = TIPO_;
  beta = real(BETA_, pr);
  kord = KORD_;
  l_interval = L_INTERVAL_; n_cuad = N_CUAD_; nev = NEV_;
  me = real(ME_, pr);
  az = real(AZ_, pr); bz = real(BZ_, pr); r0 = real(R0_, pr); rmax = real(RMAX_, pr);
  v1 = real(V1_, pr); v2 = real(V2_, pr);
  delta = real(DELTA_, pr); eta = real(ETA_, pr);
  bcampo_i = real(BCAMPO_I_, pr); bcampo_f = real(BCAMPO_F_, pr);
  num_puntos_b = NUM_PUNTOS_B_;

  nk = l_interval+2*kord-1;    ! # de knots
  ! No incluimos primer   y ultimo spline debido a psi(0)=psi(R)=0
  nb = kord+l_interval-3;      ! tamaño base splines
  ndimh = nb*(nb+1)/2;

  slope_ll = eV/(me*c_light*ua_to_T)

  write(file_auval, '("./res21032017/2e-E_vs_B-v1_",f6.4,"eV-v2_",f6.4,"eV-az_",f6.4,"-bz_",f6.4,"-r0_",f6.4,"&
  &-rmax_",f6.4,"-eta_",f8.6,"-delta_",f6.4,".dat")') v1, v2, az, bz, r0, rmax, eta, delta
  write(file_exp, '("./res21032017/2e-Int_vs_B-v1_",f6.4,"eV-v2_",f6.4,"eV-az_",f6.4,"-bz_",f6.4,"-r0_",f6.4"&
  &-rmax_",f6.4,"-eta_",f8.6,"-delta_",f6.4,".dat")') v1, v2, az, bz, r0, rmax, eta, delta
  open(10, file = file_auval)
  open(11, file = file_exp)
  write(10,'(A28,f8.2,A1,f8.2,A4)') "# Intervalo de integracion:[", zmin,",", zmax, "] nm"
  write(10,'(A32,x,I2,x,A33)') "# Tipo de distribucion de knots:", tipo, "(1 es uniforme, 2 es exponencial)"
  write(10,'(A47,x,f8.6)') "# Cte de decaimiento en distribucion exp beta =", beta
  write(10,'(A30,x,I2)') "# Orden de los bsplines kord =", kord
  write(10,'(A19,x,I3,x,A24,x,I3)') "# Num de intervalos", l_interval, ", grado de la cuadratura", n_cuad
  write(10,'(A14,x,I3,x,A24,x,I3)') "# Num de knost", nk, ", tamaño de la base nb =", nb
  write(10,'(A33,x,f8.6,x,A2)') "# Masa efectiva del electron me =", me, "UA"
  write(10,'(A22,x,f6.4,x,A2)') "# Altura del pozo V1 =", v1, "eV"
  write(10,'(A27,x,f6.4,x,A2)') "# Profundidad del pozo V2 =", v2, "eV"
  write(10,'(A22,x,f6.4,x,A9,x,f6.4,x,A2)') "# Radios del pozo az =", az, "nm y bz =", bz, "nm"
  write(10,'(A41,x,f6.4,x,A2)') "# Cte de rectificacion de Coulomb delta =", delta, "nm"
  write(10,'(A31,x,f6.4,x,A11,x,f6.4,x,A2)') "# Radios de la parte radia r0 =", r0, "nm y Rmax =", Rmax, "nm"
  write(10,'(A22,x,f6.4)') "# Carga efectiva eta =", eta
  write(10,'(A27,x,f6.2,x,A19,x,f7.2)') "# Campo inicial b_campo_i =", bcampo_i, "y final b_campo_f =", bcampo_f
  write(10,'(A24)') "# Autovalores calculados"
  call flush();

  ! Paso a unidades atomicas todo
  v1 = v1/eV; v2 = v2/eV;
  zmin = zmin/a0; zmax = zmax/a0;
  beta = beta*a0; delta = delta/a0;
  az = az/a0; bz = bz/a0; r0 = r0/a0; rmax = rmax/a0;

  allocate(k(nk), t(nk))
  allocate(x(l_interval,n_cuad), w(l_interval,n_cuad), pl(l_interval,n_cuad))
  allocate(s(nb,nb), v01(nb,nb), v02(nb,nb), ke(nb,nb))
  allocate(V_int(nb,nb,nb,nb))
  allocate(h(ndimh,ndimh), ms(ndimh,ndimh), mv(ndimh,ndimh))
  allocate(auval(nev), auvec(ndimh,nev))
  allocate(exp_int(nev,nev))

  !! calculos los knots y los puntos de la cuadratura
  call KNOTS_PESOS(kord, tipo, beta, zmin, zmax, l_interval, n_cuad, t, k, x, w, pl)

  !! ahora paso a calcular las matrices del problema
  s(:,:) = 0._pr; v01(:,:) = 0._pr; v02(:,:) = 0._pr; ke(:,:) = 0._pr;
  call calculo_matrices(kord, l_interval, n_cuad, nk, nb, me, az, bz, t, k, x, w, s, v01, v02, ke);

  call interaccion(kord, l_interval, n_cuad, nk, nb, delta, k, t, x, w, V_int)

  delta_b = (bcampo_f-bcampo_i)/real(num_puntos_b, pr);

  ind_b = 0;
  do ind_b = 0, num_puntos_b

    bcampo = bcampo_i + delta_b*real(ind_b, pr);

    l_campo = sqrt(2.0_pr*alpha/bcampo)/a0;
    omega = 0.5_pr*bcampo/(me*c_light*ua_to_T);  !! Frecuencia de oscilacion debida al campo

    vr = (1._pr - exp(-(r0/l_campo)**2))
    vrmax = (1._pr - exp(-(rmax/l_campo)**2));

    call hamiltoniano(nb, vrmax*v1, vr*v2, eta, s, v01, v02, ke, v_int, h, ms, mv);

    auval = 0._pr
    call eigenvalues(ndimh, nev, h, ms, auval, auvec)

    auval = eV*(auval + 2.0_pr*omega); !slope_ll*bcampo;

    exp_int = matmul(transpose(auvec), matmul(mv, auvec))

    write(10,6) bcampo, (auval(i), i = 1, nev)
    write(11,6) bcampo, (exp_int(i,i), i = 1, nev)
    call flush();
  end do

  deallocate(k, t)
  deallocate(x, w, pl)
  deallocate(s, v01, v02, ke)
  deallocate(h, ms)
  deallocate(auval)

  close(10)
  close(11)

6 format(e22.14,1x,1000(1x,e22.14))
! 7 format(3000(1x,e22.14))
end program
